pkg/ring/ring.go

```diff
@@ -335,20 +338,49 @@ func (r *Ring) GetAll(op Operation) (ReplicationSet, error) {
 	maxUnavailable := r.cfg.ReplicationFactor / 2
 	numRequired -= maxUnavailable
 
-	ingesters := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	maxUnavailableZones := len(r.ringZones) - maxUnavailable - 1
+	if maxUnavailableZones < 0 {
+		maxUnavailableZones = 0
+	}
+
+	instances := make([]IngesterDesc, 0, len(r.ringDesc.Ingesters))
+	zoneFailures := make(map[string]int)
 	for _, ingester := range r.ringDesc.Ingesters {
 		if r.IsHealthy(&ingester, op) {
-			ingesters = append(ingesters, ingester)
+			instances = append(instances, ingester)
+		} else {
+			zoneFailures[ingester.Zone]++
+		}
+	}
+
+	if r.cfg.ZoneAwarenessEnabled && len(zoneFailures) > 0 {
```

Comment:
Should we also add `&& maxUnavailableZones > 0`? I think we shouldn't enter this logic if we can't tolerate any zone failure.

Reply:
`maxUnavailableZones = 0` when the replication factor = 1. If any zone failures occur that will be caught by `if len(zoneFailures) > maxUnavailableZones` on the next line. So I don't think it's absolutely necessary.

On the other hand it won't hurt so it's been added. I've added extra test cases that have a replication factor of 1 and 1 or more zones. These tests up the coverage as well.